### ISSUE TITLE
UploadFromByteArrayAsync "leaks" handles

### ISSUE THREAD

**Issue by benaadams** [Azure/azure-storage-net#224]
Calling UploadFromByteArrayAsync repeatedly on coreclr x64 continuously increases handles in use and number of open sockets

Repo: https://github.com/benaadams/TestCase-AzureStorageHandles/blob/master/src/AzureStorageHandles/Program.cs

![handles-1](http://aoa.blob.core.windows.net/aspnet/handles1.png)
Few minutes later
![handles-2](http://aoa.blob.core.windows.net/aspnet/handles2.png)

![tcp-connections](http://aoa.blob.core.windows.net/aspnet/connections.png)

**Comment by benaadams** [comment]
At a guess this is due to setting 
```request.KeepAlive = true;``` 
In `HttpWebRequestFactory`.`CreateWebRequest`
https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Shared/Protocol/HttpWebRequestFactory.cs#L80
Even though the created `HttpWebRequest`s are never reused but created each time.

**Comment by benaadams** [comment]
Left running for a long while the handles periodically flush; though they keep growing and memory use climbs (likely due to lake of GC pressure) - though it has the appearances of a leak

![handles](http://aoa.blob.core.windows.net/aspnet/handles3.png)

Commenting out the upload line cause this all to disappear

**Comment by mirobers** [comment]
Thanks for reporting this. We are investigating.

**Comment by pemari-msft** [comment]
Hi @benaadams, I took a look at this. To clarify, KeepAlive shouldn't be related. CoreCLR uses the HttpClient & request message paradigm and not the HttpWebRequest paradigm, and in fact does not use that file when compiling.

I do however see that we aren't reusing connections in our CoreCLR code. We'll file a bug and look to fix this in an upcoming release. However, I did not see nearly as bad of a repro using your repro code on my machine. I never saw the handles increase above 1100 and I saw them regularly flushing even after an hour, so I wonder what the specs are on your repro machine? In any case, thanks for reporting this!
